clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRToMLIR.cpp

```diff
@@ -468,30 +468,26 @@ class CIRSinOpLowering : public mlir::OpConversionPattern<cir::SinOp> {
 };
 
 template <typename CIROp, typename MLIROp>
-class CIRBitOpLowering : public mlir::OpConversionPattern<CIROp> {
+class CIRCountZerosBitOpLowering : public mlir::OpConversionPattern<CIROp> {
 public:
   using mlir::OpConversionPattern<CIROp>::OpConversionPattern;
 
   mlir::LogicalResult
   matchAndRewrite(CIROp op,
                   typename mlir::OpConversionPattern<CIROp>::OpAdaptor adaptor,
                   mlir::ConversionPatternRewriter &rewriter) const override {
-    auto resultIntTy = mlir::cast<mlir::IntegerType>(
-        this->getTypeConverter()->convertType(op.getType()));
-    auto res = rewriter.create<MLIROp>(op->getLoc(), adaptor.getInput());
-    auto newOp = createIntCast(rewriter, res->getResult(0), resultIntTy,
-                               /*isSigned=*/false);
-    rewriter.replaceOp(op, newOp);
+    if (op.getIsZeroPoison()) {
+      return mlir::LogicalResult::failure();
```

Comment:
Please emit a short error message before returning failure.

Reply:
> Other Operation's matchAndRewrite which returns failure doesn't emit an error message. Is there any code I can take as reference?

You can emit a diagnostic message via [`op.emitError` function](https://mlir.llvm.org/doxygen/classmlir_1_1OpState.html#a78ab2e156408bc77bb5d5629068bda60). Maybe you could refer to `LowerToLLVM.cpp` instead for examples; there are quite a few such calls.